File: 2015/2015.clj
;;;;
;;;;
;;;;   Clojure feels like a general-purpose language beamed back from the near future.
;;;;   -- Stu Halloway
;;;;
;;;;   Name:               2015.clj
;;;;
;;;;   Started:            Fri Sep 10 12:25:18 2021
;;;;
;;;;   Purpose:
;;;;   Purpose: Advent of Code 2015
;;;;   https://adventofcode.com/2015
;;;;
;;;;   Notes:
;;;;
;;;;

(ns two-thousand-fifteen
  (:use clojure.test
        [clojure.pprint :only (cl-format)])
  (:require [clojure.string :as string])
  (:import [java.security MessageDigest]))

;;;
;;;    Day 1
;;;
(defn deliver-presents
  "Determine floor for Santa to deliver presents based on instructions `s`. `(` means up one floor. `)` means down one floor."
  [s]
  (reduce (fn [floor ch]
            (case ch
              \( (inc floor)
              \) (dec floor)))
          0
          s))

(defn deliver-presents
  "Determine floor for Santa to deliver presents based on instructions `s`. `(` means up one floor. `)` means down one floor."
  [s]
  (loop [i 0
         floor 0]
    (if (== i (count s))
      floor
      (case (get s i)
        \( (recur (inc i) (inc floor))
        \) (recur (inc i) (dec floor)))) ))

(deftest test-deliver-presents ()
  (is (== (deliver-presents "(())") 0))
  (is (== (deliver-presents "()()") 0))
  (is (== (deliver-presents "(((") 3))
  (is (== (deliver-presents "(()(()(") 3))
  (is (== (deliver-presents "))(((((") 3))
  (is (== (deliver-presents "())") -1))
  (is (== (deliver-presents "))(") -1))
  (is (== (deliver-presents ")))") -3))
  (is (== (deliver-presents ")())())") -3)))

;;;
;;;    At first, it seems more natural to treat the string as a sequence,
;;;    but there is some awkwardness.
;;;    
(defn enter-basement
  "Determine when (or whether) Santa enters the basement. 1-based index."
  [s]
  (loop [ch (first s)
         s (rest s)
         floor (case ch \( 1 \) -1)
         index 1]
    (cond (== floor -1) index
          (empty? s) false
          :else (case (first s)
                  \( (recur (first s) (rest s) (inc floor) (inc index))
                  \) (recur (first s) (rest s) (dec floor) (inc index)))) ))

(defn enter-basement
  "Determine when (or whether) Santa enters the basement. 1-based index."
  [s]
  (loop [i 0
         floor 0]
    (cond (== floor -1) i
          (== i (count s)) false
          :else (case (get s i)
                  \( (recur (inc i) (inc floor))
                  \) (recur (inc i) (dec floor)))) ))

(deftest test-enter-basement ()
  (is (== (enter-basement ")") 1))
  (is (== (enter-basement "()())") 5))
  (is (not (enter-basement "()"))))

;;;
;;;    Day 2
;;;    
(defn compute-wrap [length width height]
  (let [[a b c] (sort [length width height])
        ab (* a b)
        ac (* a c)
        bc (* b c)
        slack ab]
    (+ (* 2 ab) (* 2 ac) (* 2 bc) slack)))

(deftest test-compute-wrap ()
  (is (== (compute-wrap 2 3 4) 58))
  (is (== (compute-wrap 1 1 10) 43)))

(defn parse-dimensions [dimensions]
  (map read-string (string/split dimensions #"x")))

;;;
;;;    Don't `map` twice!
;;;    
;; (defn calculate-total-wrap [file]
;;   (reduce + (map #(apply compute-wrap %) (map parse-dimensions (string/split (slurp file) #"\n")))) )

(defn calculate-total-wrap [file]
  (reduce + (map #(apply compute-wrap (parse-dimensions %)) (string/split (slurp file) #"\n"))))

(defn calculate-total-quantity [file compute-quantity]
  (reduce + (map (comp (partial apply compute-quantity) parse-dimensions) (string/split (slurp file) #"\n"))))

(defn calculate-total-wrap [file]
  (calculate-total-quantity file compute-wrap))

(defn compute-ribbon [length width height]
  (let [[a b c] (sort [length width height])
        ribbon (+ (* 2 a) (* 2 b))
        bow (* a b c)]
    (+ ribbon bow)))

(deftest test-compute-ribbon ()
  (is (== (compute-ribbon 2 3 4) 34))
  (is (== (compute-ribbon 1 1 10) 14)))

;;;
;;;    Don't `map` twice!
;;;    
;; (defn calculate-total-ribbon [file]
;;   (reduce + (map #(apply compute-ribbon %) (map parse-dimensions (string/split (slurp file) #"\n")))) )

(defn calculate-total-ribbon [file]
  (reduce + (map #(apply compute-ribbon (parse-dimensions %)) (string/split (slurp file) #"\n"))))

(defn calculate-total-ribbon [file]
  (calculate-total-quantity file compute-ribbon))

;;;
;;;    Day 3
;;;    
(defn visit-houses [s]
  (loop [s s
         x 0
         y 0
         visited {[x y] 1}]
    (cond (empty? s) (count visited)
          :else (let [ch (first s)
                      x1 (case ch \< (dec x) \> (inc x) x)
                      y1 (case ch \^ (inc y) \v (dec y) y)]
                  (recur (rest s) x1 y1 (assoc visited [x1 y1] (inc (visited [x1 y1] 0)))) ))))

(defn visit-houses [s]
  (let [leave-present (fn [visited x y]
                        (assoc visited [x y] (inc (visited [x y] 0))))
        update-location (fn [x y ch]
                          [(case ch \< (dec x) \> (inc x) x)
                           (case ch \^ (inc y) \v (dec y) y)])]
    (loop [s s
           x 0
           y 0
           visited {}]
      (let [visited (leave-present visited x y)]
        (cond (empty? s) (count visited)
              :else (let [[x1 y1] (update-location x y (first s))]
                      (recur (rest s) x1 y1 visited)))) )))

(defn visit-houses [s]
  (letfn [(leave-present [visited x y]
            (let [current (visited [x y] 0)]
              (assoc visited [x y] (inc current))))
          (update-location [x y ch]
            [(case ch \< (dec x) \> (inc x) x)
             (case ch \^ (inc y) \v (dec y) y)])
          (visit [s x y visited]
            (let [new-visited (leave-present visited x y)]
              (cond (empty? s) (count new-visited)
                    :else (let [[x1 y1] (update-location x y (first s))]
                            (recur (rest s) x1 y1 new-visited)))) )]
    (visit s 0 0 {})))

(deftest test-visit-houses ()
  (is (== (visit-houses ">") 2))
  (is (== (visit-houses "^>v<") 4))
  (is (== (visit-houses "^v^v^v^v^v") 2)))

;(visit-houses (slurp "/home/slytobias/lisp/books/Advent/advent/2015/day3.data")) => 2081

(defrecord Agent [x y])

(defn make-agent
  ([] (make-agent 0 0))
  ([x y] (Agent. x y)))

(defn location [agent]
  [(:x agent) (:y agent)])

(defn leave-present
  "Leave a present at the agent's location"
  [agent visited]
  (let [loc (location agent)]
    (assoc visited loc (inc (visited loc 0)))) )

(defn update-location
  "Update the agent's location in the given direction."
  [agent direction]
  (let [x (:x agent)
        y (:y agent)
        x1 (case direction \< (dec x) \> (inc x) x)
        y1 (case direction  \^ (inc y) \v (dec y) y)]
    (make-agent x1 y1)))

(def update-map {\< (fn [[x y]] [(dec x) y])
                 \> (fn [[x y]] [(inc x) y])
                 \^ (fn [[x y]] [x (inc y)])
                 \v (fn [[x y]] [x (dec y)])})
(defn update-location
  "Update the agent's location in the given direction."
  [agent direction]
  (let [[x1 y1] ((update-map direction) (location agent))]
    (make-agent x1 y1)))

;; (defn robo-visit [s]
;;   (let [santa (make-agent)
;;         robot (make-agent)
;;         visited (leave-present robot (leave-present santa {}))]
;;     (loop [q (conj (conj clojure.lang.PersistentQueue/EMPTY santa) robot)
;;            s s
;;            visited visited]
;;       (cond (empty? s) (count visited)
;;             :else (let [agent (update-location (peek q) (first s))
;;                         new-q (pop q)]
;;                     (recur (conj new-q agent) (rest s) (leave-present agent visited)))) )))

(defn robo-visit [s]
  (let [santa (make-agent)
        robot (make-agent)
        visited (leave-present robot (leave-present santa {}))
        team (conj (conj clojure.lang.PersistentQueue/EMPTY santa) robot)]
    (letfn [(visit [s team visited]
              (cond (empty? s) (count visited)
                    :else (let [agent (update-location (peek team) (first s))
                                new-team (pop team)]
                            (recur (rest s) (conj new-team agent) (leave-present agent visited)))) )]
      (visit s team visited))))

(deftest test-robo-visit ()
  (is (== (robo-visit "^v") 3))
  (is (== (robo-visit "^>v<") 3))
  (is (== (robo-visit "^v^v^v^v^v") 11)))
   
;; (robo-visit (slurp "/home/slytobias/lisp/books/Advent/advent/2015/day3.data")) => 2341

;;;
;;;    Day 4
;;;    
;;;
;;;    Inspired by:
;;;    https://gist.github.com/jizhang/4325757
;;;    Aleksander Madland Stapnes
(defn coin-found? [^String seed n]
  (let [nybbles (mapcat #(vector (quot % 16) (rem % 16))
                        (map #(Byte/toUnsignedInt %) (vec (.digest (MessageDigest/getInstance "MD5") (.getBytes seed)))) )]
    (every? zero? (take n nybbles))))

(defn mine-advent-coin [^String prefix n]
  (loop [i 1]
    (let [seed (cl-format false "~A~D" prefix i)]
      (if (coin-found? seed n)
        i
        (recur (inc i)))) ))

;;;
;;;    This is very slow! Need to figure out how to speed up...
;;;    
;; (deftest test-mine-advent-coin ()
;;   (is (== (mine-advent-coin "abcdef" 5) 609043))
;;   (is (== (mine-advent-coin "pqrstuv" 5) 1048970)))

;; (mine-advent-coin "ckczppom" 5) => 117946
;; (mine-advent-coin "ckczppom" 6) => 3938038

;;;
;;;    Day 5
;;;    
(defn nice-string? [s]
  (letfn [(nice? [vowels double?]
            (and (>= vowels 3) double?))
          (vowel? [ch]
            (contains? #{\a \e \i \o \u} ch))
          (check-double [ch s vowels double?]
            (let [next-ch (first s)]
              (cond (nil? next-ch) (nice? vowels double?)
                    :else (check-nice next-ch (rest s) vowels (or (= ch next-ch) double?)))) )
          (check-nice [ch s vowels double?]
            (cond (nil? ch) (nice? vowels double?)
                  :else (case ch
                          \a (check-a (first s) (rest s) (inc vowels) double?)
                          \c (check-c (first s) (rest s) vowels double?)
                          \p (check-p (first s) (rest s) vowels double?)
                          \x (check-x (first s) (rest s) vowels double?)
                          (check-double ch s (if (vowel? ch) (inc vowels) vowels) double?))))
          (check-a [ch s vowels double?]
            (cond (nil? ch) (nice? vowels double?)
                  :else (case ch
                          \b false
                          \a (check-a (first s) (rest s) (inc vowels) true)
                          \c (check-c (first s) (rest s) vowels double?)
                          \p (check-p (first s) (rest s) vowels double?)
                          \x (check-x (first s) (rest s) vowels double?)
                          (check-double ch s (if (vowel? ch) (inc vowels) vowels) double?))))
          (check-c [ch s vowels double?]
            (cond (nil? ch) (nice? vowels double?)
                  :else (case ch
                          \d false
                          \a (check-a (first s) (rest s) (inc vowels) double?)
                          \c (check-c (first s) (rest s) vowels true)
                          \p (check-p (first s) (rest s) vowels double?)
                          \x (check-x (first s) (rest s) vowels double?)
                          (check-double ch s (if (vowel? ch) (inc vowels) vowels) double?))))
          (check-p [ch s vowels double?]
            (cond (nil? ch) (nice? vowels double?)
                  :else (case ch
                          \q false
                          \a (check-a (first s) (rest s) (inc vowels) double?)
                          \c (check-c (first s) (rest s) vowels double?)
                          \p (check-p (first s) (rest s) vowels true)
                          \x (check-x (first s) (rest s) vowels double?)
                          (check-double ch s (if (vowel? ch) (inc vowels) vowels) double?))))
          (check-x [ch s vowels double?]
            (cond (nil? ch) (nice? vowels double?)
                  :else (case ch
                          \y false
                          \a (check-a (first s) (rest s) (inc vowels) double?)
                          \c (check-c (first s) (rest s) vowels double?)
                          \p (check-p (first s) (rest s) vowels double?)
                          \x (check-x (first s) (rest s) vowels true)
                          (check-double ch s (if (vowel? ch) (inc vowels) vowels) double?)))) ]
    (check-nice (first s) (rest s) 0 false)))

(deftest test-nice-string? ()
  (is (nice-string? "ugknbfddgicrmopn"))
  (is (nice-string? "aaa"))
  (is (not (nice-string? "jchzalrnumimnmhp")))
  (is (not (nice-string? "haegwjzuvuyypxyu")))
  (is (not (nice-string? "dvszwmarrgswjxmb"))))

;; (count (remove false? (map nice-string? (string/split (slurp "day5.data") #"\n")))) => 236

(defn nice-string?* [s]
  (let [rule1 (re-find #"(.)(.).*\1\2" s)
        rule2 (re-find #"(.).\1" s)]
    (if (and rule1 rule2)
      true
      false)))

(deftest test-nice-string?* ()
  (is (nice-string?* "qjhvhtzxzqqjkmpb"))
  (is (nice-string?* "xxyxx"))
  (is (nice-string?* "qryjbohkprfazczc")) ; Check index for Rule 1!
  (is (not (nice-string?* "uurcxstgmygtbstg")))
  (is (not (nice-string?* "ieodomkazucvgmuy")))
  (is (not (nice-string?* "suerykeptdsutidb")))) ; Check index for Rule 2!

;; (count (remove false? (map nice-string?* (string/split (slurp "day5.data") #"\n")))) => 51

File: 2015/2015.pl
#!/usr/bin/swipl -q -t main -f
%%
%   -*- Mode: Prolog -*-
%   Name:               2015.pl
%
%   Started:            Fri Sep 10 20:07:50 2021
%
%   Purpose: Advent of Code 2015
%   https://adventofcode.com/2015
%
%   Notes:
%
%%

:- module(advent_2015, [deliver_presents/2, enter_basement/2, compute_wrap/2, compute_ribbon/2]).

:- use_module('/home/slytobias/prolog/books/Adventure/io.pl').

%%%
%%%    Day 1
%%%    
deliver_presents(S, Floor) :-
    atom_chars(S, L),
    deliver_presents(L, Floor, 0).
deliver_presents([], Floor, Floor).
deliver_presents(['('|T], Floor, Acc) :-
    Acc1 is Acc + 1,
    deliver_presents(T, Floor, Acc1).
deliver_presents([')'|T], Floor, Acc) :-
    Acc1 is Acc - 1,
    deliver_presents(T, Floor, Acc1).

%% deliver_presents('(())', 0).
%% deliver_presents('()()', 0).
%% deliver_presents('(((', 3).
%% deliver_presents('(()(()(', 3).
%% deliver_presents('))(((((', 3).
%% deliver_presents('())', -1).
%% deliver_presents('))(', -1).
%% deliver_presents(')))', -3).
%% deliver_presents(')())())', -3).

start_floor('(', 1).
start_floor(')', -1).

enter_basement(S, Index) :-
    atom_chars(S, [H|T]),
    start_floor(H, F),
    enter_basement(T, F, 1, Index).
enter_basement(_, -1, I, I).
enter_basement(['('|T], F, I, Index) :-
    F1 is F + 1,
    I1 is I + 1,
    enter_basement(T, F1, I1, Index).
enter_basement([')'|T], F, I, Index) :-
    F1 is F - 1,
    I1 is I + 1,
    enter_basement(T, F1, I1, Index).

%% enter_basement(')', 1).
%% enter_basement('()())', 5).
%% enter_basement('()', F).

%%%
%%%    Day 2
%%%
compute_wrap([Length, Width, Height], Wrap) :-
    msort([Length, Width, Height], [A, B, C]), % Don't remove duplicates!!
    AB is A * B,
    AC is A * C,
    BC is B * C,
    Slack is AB,
    Wrap is 2 * AB + 2 * AC + 2 * BC + Slack.

%% compute_wrap(2, 3, 4, 58).
%% compute_wrap(1, 1, 10, 43).

parse_dimensions(D, Ns) :-
    atom_string(D, S),
    split_string(S, "x", "", Ds),
    convert_dimensions(Ds, Ns).

convert_dimensions([], []).
convert_dimensions([D|Ds], [N|Ns]) :-
    number_string(N, D),
    convert_dimensions(Ds, Ns).

calculate_total_wrap(File, Sum) :-
    io:read_file(File, Lines),
    parse_file(Lines, Dimensions),
    compute_all_wraps(Dimensions, Sum).

compute_all_wraps([], 0).
compute_all_wraps([D|Ds], S) :-
    compute_wrap(D, W),
    compute_all_wraps(Ds, S1),
    S is S1 + W.

parse_file([], []).
parse_file([L|Ls], [D|Ds]) :-
    parse_dimensions(L, D),
    parse_file(Ls, Ds).

compute_ribbon([Length, Width, Height], R) :-
    msort([Length, Width, Height], [A, B, C]),
    Ribbon is 2 * A + 2 * B,
    Bow is A * B * C,
    R is Ribbon + Bow.

%% compute_ribbon(2, 3, 4, 34).
%% compute_ribbon(1, 1, 10, 14).

%% (defn calculate-total-ribbon [file]
%%   (reduce + (map #(apply compute-ribbon %) (map parse-dimensions (string/split (slurp file) #"\n")))) )
calculate_total_ribbon(File, Sum) :-
    io:read_file(File, Lines),
    parse_file(Lines, Dimensions),
    compute_all_ribbons(Dimensions, Sum).

compute_all_ribbons([], 0).
compute_all_ribbons([D|Ds], S) :-
    compute_ribbon(D, R),
    compute_all_ribbons(Ds, S1),
    S is S1 + R.

%%%
%%%    Day 3
%%%    Probably not the right way to do this...
%%%    Dynamically uses KB to record visited locations.
%%%    Should use SWIPL dictionary instead??
%%%
:- dynamic visited/3.

update_location('^', X, Y, X, Y1) :-
    Y1 is Y + 1.
update_location('v', X, Y, X, Y1) :-
    Y1 is Y - 1.
update_location('>', X, Y, X1, Y) :-
    X1 is X + 1.
update_location('<', X, Y, X1, Y) :-
    X1 is X - 1.

leave_present(X, Y) :-
    visited(X, Y, C), !,
    retract(visited(X, Y, C)),
    C1 is C + 1,
    assertz(visited(X, Y, C1)).
leave_present(X, Y) :-
    \+ visited(X, Y, _),
    assertz(visited(X, Y, 1)).

%%%
%%%    Nonsense!!!
%%%    
%% visit_houses(Directions, Count) :-
%% %    retract(visited(_, _, _)),
%%     retractall(visited(_, _, _)),
%%     leave_present(0, 0),
%%     visit_houses(Directions, Count, 1, 0, 0).
%% visit_houses([], Count, Count, _, _) :- !.
%% visit_houses([D|Ds], Count, C, X, Y) :-
%%     update_location(D, X, Y, X1, Y1),
%%     leave_present(X1, Y1),
%%     C1 is C + 1,
%%     visit_houses(Ds, Count, C1, X1, Y1).

%% visit_houses(Directions, Count) :-
%% %    retract(visited(_, _, _)),
%%     retractall(visited(_, _, _)),
%%     leave_present(0, 0),
%%     visit_houses(Directions, Count, 0, 0).
%% visit_houses([], Count, _, _) :-
%%     findall(C, visited(_, _, C), L),
%%     length(L, Count).
%% visit_houses([D|Ds], Count, X, Y) :-
%%     update_location(D, X, Y, X1, Y1),
%%     leave_present(X1, Y1),
%%     visit_houses(Ds, Count, X1, Y1).

visit_houses(Directions, Count) :-
%    retract(visited(_, _, _)),
    retractall(visited(_, _, _)),
    visit_houses(Directions, Count, 0, 0).
visit_houses([], Count, X, Y) :-
    leave_present(X, Y),
    findall(C, visited(_, _, C), L),
    length(L, Count).
visit_houses([D|Ds], Count, X, Y) :-
    leave_present(X, Y),
    update_location(D, X, Y, X1, Y1),
    visit_houses(Ds, Count, X1, Y1).

%% io:read_chars('day3.data', L), visit_houses(L, C).
%% L = [^, >, <, ^, >, >, >, ^, <|...],
%% C = 2081 

leave_present(agent(_, X, Y)) :-
    leave_present(X, Y).

update_location(agent(Name, X, Y), Dir, agent(Name, X1, Y1)) :-
    update_location(Dir, X, Y, X1, Y1).

robo_visit(Directions, Count) :-
    retractall(visited(_, _, _)),
    Santa = agent(santa, 0, 0),
    Robot = agent(robot, 0, 0),
    leave_present(Santa),
    leave_present(Robot),
    Team = [Santa, Robot],
    robo_visit(Directions, Team, Count).
robo_visit([], _, Count) :-
    findall(C, visited(_, _, C), L),
    length(L, Count).
robo_visit([D|Ds], [A|As], Count) :-
    update_location(A, D, A1),
    leave_present(A1),
    append(As, [A1], Team),
    robo_visit(Ds, Team, Count).
    
%% io:read_chars('day3.data', L), robo_visit(L, C).
%% L = [^, >, <, ^, >, >, >, ^, <|...],
%% C = 2341 

%%%
%%%    Day 4
%%%    
:- use_module(library(md5)).

mine_advent_coin(Prefix, N, Coin) :-
    mine_advent_coin(Prefix, N, 1, Coin).
mine_advent_coin(Prefix, N, I, I) :-
    atom_concat(Prefix, I, Seed),
    coin_found(Seed, N).
mine_advent_coin(Prefix, N, I, Coin) :-
    I1 is I + 1,
    mine_advent_coin(Prefix, N, I1, Coin).

coin_found(Seed, N) :-
    md5_hash(Seed, M, []),
    atom_chars(M, Nybbles),
    first_n_zeroes(Nybbles, N).

first_n_zeroes(_, 0).
first_n_zeroes(['0'|As], N) :-
    N1 is N - 1,
    first_n_zeroes(As, N1).

%% advent_2015: 106 ?- mine_advent_coin("ckczppom", 5, I).
%% I = 117946

%% advent_2015: 107 ?- mine_advent_coin("ckczppom", 6, I).
%% I = 3938038 

%%%
%%%    Day 5
%%%    So many cuts!!
%%%    
check_nice1(File, Count) :-
    io:read_file(File, Lines),
    count_nice1(Lines, Count, 0).

count_nice1([], Count, Count).
count_nice1([L|Ls], Count, C) :-
    nice1(L),
    C1 is C + 1,
    count_nice1(Ls, Count, C1).
count_nice1([_|Ls], Count, C) :-
    count_nice1(Ls, Count, C).

nice1(S) :-
    atom_chars(S, L),
    check_nice(L, 0, false).

nice(V, true) :-
    V >= 3.

vowel(a).
vowel(e).
vowel(i).
vowel(o).
vowel(u).

check_nice([], V, D) :-
    nice(V, D).
check_nice([a|Cs], V, D) :-
    !,
    V1 is V + 1,
    check_a(Cs, V1, D).
check_nice([c|Cs], V, D) :-
    !, check_c(Cs, V, D).
check_nice([p|Cs], V, D) :-
    !, check_p(Cs, V, D).
check_nice([x|Cs], V, D) :-
    !, check_x(Cs, V, D).
check_nice([C|Cs], V, D) :-
    vowel(C), !,
    V1 is V + 1,
    check_double_nice(C, Cs, V1, D).
check_nice([C|Cs], V, D) :-
    check_double_nice(C, Cs, V, D).

check_double_nice(_, [], V, D) :-
    nice(V, D).
check_double_nice(A, [A|Cs], V, _) :-
    !,
    check_nice([A|Cs], V, true).
check_double_nice(_, [A|Cs], V, D) :-
    check_nice([A|Cs], V, D).
    
check_a([], V, D) :-
    nice(V, D).
check_a([b|_], _, _) :- !, fail.
check_a([a|Cs], V, _) :-
    !,
    V1 is V + 1,
    check_a(Cs, V1, true).
check_a([c|Cs], V, D) :-
    !, check_c(Cs, V, D).
check_a([p|Cs], V, D) :-
    !, check_p(Cs, V, D).
check_a([x|Cs], V, D) :-
    !, check_x(Cs, V, D).
check_a([C|Cs], V, D) :-
    vowel(C), !,
    V1 is V + 1,
    check_double_nice(C, Cs, V1, D).
check_a([C|Cs], V, D) :-
    check_double_nice(C, Cs, V, D).

check_c([], V, D) :-
    nice(V, D).
check_c([d|_], _, _) :- !, fail.
check_c([c|Cs], V, _) :-
    !, check_c(Cs, V, true).
check_c([a|Cs], V, D) :-
    !,
    V1 is V + 1,
    check_a(Cs, V1, D).
check_c([p|Cs], V, D) :-
    !, check_p(Cs, V, D).
check_c([x|Cs], V, D) :-
    !, check_x(Cs, V, D).
check_c([C|Cs], V, D) :-
    vowel(C), !,
    V1 is V + 1,
    check_double_nice(C, Cs, V1, D).
check_c([C|Cs], V, D) :-
    check_double_nice(C, Cs, V, D).

check_p([], V, D) :-
    nice(V, D).
check_p([q|_], _, _) :- !, fail.
check_p([p|Cs], V, _) :-
    !, check_p(Cs, V, true).
check_p([a|Cs], V, D) :-
    !,
    V1 is V + 1,
    check_a(Cs, V1, D).
check_p([c|Cs], V, D) :-
    !, check_c(Cs, V, D).
check_p([x|Cs], V, D) :-
    !, check_x(Cs, V, D).
check_p([C|Cs], V, D) :-
    vowel(C), !,
    V1 is V + 1,
    check_double_nice(C, Cs, V1, D).
check_p([C|Cs], V, D) :-
    check_double_nice(C, Cs, V, D).

check_x([], V, D) :-
    nice(V, D).
check_x([y|_], _, _) :- !, fail.
check_x([x|Cs], V, _) :-
    !, check_x(Cs, V, true).
check_x([a|Cs], V, D) :-
    !,
    V1 is V + 1,
    check_a(Cs, V1, D).
check_x([c|Cs], V, D) :-
    !, check_c(Cs, V, D).
check_x([p|Cs], V, D) :-
    !, check_p(Cs, V, D).
check_x([C|Cs], V, D) :-
    vowel(C), !,
    V1 is V + 1,
    check_double_nice(C, Cs, V1, D).
check_x([C|Cs], V, D) :-
    check_double_nice(C, Cs, V, D).

%% (deftest test-nice-string-p ()
%%   (check
%%    (nice-string-p "ugknbfddgicrmopn")
%%    (nice-string-p "aaa")
%%    (not (nice-string-p "jchzalrnumimnmhp"))
%%    (not (nice-string-p "haegwjzuvuyypxyu"))
%%    (not (nice-string-p "dvszwmarrgswjxmb"))))







check_nice2(File, Count) :-
    io:read_file(File, Lines),
    count_nice2(Lines, Count, 0).

count_nice2([], Count, Count).
count_nice2([L|Ls], Count, C) :-
    nice2(L),
    C1 is C + 1,
    count_nice2(Ls, Count, C1).
count_nice2([_|Ls], Count, C) :-
    count_nice2(Ls, Count, C).

nice2(S) :-
    atom_chars(S, [A, B|Cs]),
    rule1(A, B, Cs),
    rule2(A, B, Cs).

rule1(A, B, Cs) :-
    search(A, B, Cs).
rule1(_, A, [B|Cs]) :-
    rule1(A, B, Cs).
search(A, B, [A, B|_]).
search(A, B, [_|Cs]) :-
    search(A, B, Cs).

rule2(A, _, [A|_]).
rule2(_, A, [B|Cs]) :-
    rule2(A, B, Cs).

%% advent_2015: 20 ?- nice2(qjhvhtzxzqqjkmpb).
%% true ;
%% true ;
%% false.

%% advent_2015: 21 ?- nice2(xxyxx).
%% true ;
%% false.

%% advent_2015: 22 ?- nice2(qryjbohkprfazczc).
%% true ;
%% true ;
%% false.

%% advent_2015: 23 ?- nice2(uurcxstgmygtbstg).
%% false.

%% advent_2015: 24 ?- nice2(ieodomkazucvgmuy).
%% false.

%% advent_2015: 25 ?- nice2(suerykeptdsutidb).
%% false.

%% advent_2015: 30 ?- check_nice2('day5.data', C).
%% C = 51
